### PR TITLE
add pinned messages migration

### PR DESCRIPTION
**Pull request by grvn-ht**
New Step to process all messages, get pinned one and put them to synapse

### CODE REVIEW

File: src/handlers/pinnedMessages.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+export type PinnedMessages = { [key: string]: string[] }
+export type UserForRoom = { [key: string]: string }
+export type PromiseResult = {
+    pinnedMessages: PinnedMessages;
+    userForRoom: UserForRoom;
+};
+
+/**
+ * Type of Rocket.Chat messages
```
Comment (HerHde) [inline src/handlers/pinnedMessages.ts:18]: I think it'd make more sense to import the type from `src/handlers/messages.ts` or put it into it's own definition file and import it from there. Or did you change anything?

File: src/handlers/pinnedMessages.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        log.warn(
+          `Room ${message.rid} has no mapping, thus no pinned messages.`
+        )
+        continue
+      }
+
+      if (message.pinnedBy !== undefined) {
+        log.info(matrixMessageId)
+        log.info(matrixRoomId)
+        log.info(message.pinnedBy._id)
```
Comment (HerHde) [inline src/handlers/pinnedMessages.ts:76]: Here I'd prefer one log message with a bit of context for the user

File: src/handlers/pinnedMessages.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+      const matrixRoomId = await getRoomId(message.rid)
+      const matrixMessageId = await getMessageId(message._id)
+      if (!matrixRoomId) {
+        log.warn(
+          `Room ${message.rid} has no mapping, thus no pinned messages.`
+        )
+        continue
+      }
+
+      if (message.pinnedBy !== undefined) {
```
Comment (HerHde) [inline src/handlers/pinnedMessages.ts:73]: I'd check for the existence and truthy values like

```suggestion
      if (message.pinnedBy) {
```

as well as in the other checks like this. Or do you do it intentionally like this?

File: src/handlers/pinnedMessages.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+  for (const room in userForRoom) {
+    log.info("room")
+    log.info(room)
+
+    log.info("user")
+    log.info(userForRoom[room])
+    const userSessionOptions = formatUserSessionOptions(
+      (await getMapping(userForRoom[room],entities[Entity.Users].mappingType))?.accessToken || ''
+    )
```
Comment (HerHde) [inline src/handlers/pinnedMessages.ts:111]: You could use `getUserSessionOptions(rcId: string)` instead:
```suggestion
    const userSessionOptions = getUserSessionOptions(userForRoom[room]) || ''
```

File: src/handlers/pinnedMessages.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  promiseResult: PromiseResult
+) {
+  const { userForRoom } = promiseResult;
+
+  for (const room in userForRoom) {
+    log.info("room")
+    log.info(room)
+
+    log.info("user")
+    log.info(userForRoom[room])
```
Comment (HerHde) [inline src/handlers/pinnedMessages.ts:108]: Also here a more informative and user-friendly log would be appreciated

**Review by HerHde — COMMENTED** [review]
Well done, thank you! I added some remarks, but in general it looks good to me, without testing the functionality, yet.

Would you mind to provide tests and/or maybe some dummy data for me to test it with?

File: src/handlers/pinnedMessages.ts (hunk 1 above)
Comment (grvn-ht) [reply src/handlers/pinnedMessages.ts:18]: i added 
  pinnedBy?: {
    _id?: string
    username?: string
  }
But i could add it to messages.ts directly

File: src/handlers/pinnedMessages.ts (hunk 3 above)
Comment (grvn-ht) [reply src/handlers/pinnedMessages.ts:73]: no I think your way is fine!

**Comment by grvn-ht** [comment]
Thanks for reviewing, going to make changes!

**Comment by grvn-ht** [comment]
Hi @HerHde ,
I made changes you suggested, but I'm struggling a bit more with tests implementation.
To try it with dummed data I would suggest to pinn a message in rocketchat server and then export the data but I don't think it's what you were thinking of.
What would you suggest?

**Comment by HerHde** [comment]
I'm working on some test data (although I'm barely at work, currently), however, I was thinking more about unit tests.

As I don't know your experience on the matter and for clarity: It would be helpful to have tests for each function that test if they do what they say. So `getPinnedMessages()` should get pinned messages from an input file and return the `PromiseResult` with valid data, but for the test we'd mock the `lineByLine` and other *side effects* to provide some test data, which can be a minimal version of what we need.
Other side effects, like requests to Axios can the *expected* to `toHaveBeenCalledWith` specific arguments, testing that the request is made as expected.

There are [some examples of such tests](https://github.com/verdigado/rocketchat2matrix/blob/b60454e81e7a90ea211dafa8d87f9a660ec6cd62/src/handlers/messages.test.ts#L64), without the specific detail of reading a file, but similar mocks.

Feel free to try out. If you need help, I'll be back the week after next. And in the thick of it, I can do it.

**Comment by HerHde** [comment]
The current main branch (at f99771b2803ccc49dae3c28e280fadee96adb212) contains some test data, including three pinned messages. You might want to rebase onto that.

**Comment by grvn-ht** [comment]
Hey @HerHde ,
thank's i will try

**Comment by HerHde** [comment]
> So I need to execute a script to change power level in all rooms before executing this migration.

Just an idea, could we maybe use the admin for pinning? Could you try to move `removeExcessRoomMembers()` to the end of the commands in `app.ts` and remove `userSessionOptions` from `axios.put`. For a `User ... not in room` error we could wrap the request in `executeAndHandleMissingMember()`.

**Comment by HerHde** [comment]
Okay, I tired it in 56911bb995921bb18a231cf9752579c7174a0a68 and had a couple of problems:

- I couldn't add an admin to the room which was not in the RC data as it has no mapping
- After adding the admin manually, I still got the same error
  
  ```json
  {
	"errcode": "M_FORBIDDEN",
	"error": "You don't have permission to post that to the room. user_level (0) < send_level (50)"
  }
  ```

**Comment by grvn-ht** [comment]
> Just an idea, could we maybe use the admin for pinning?

I think you're right! this will work

**Comment by grvn-ht** [comment]
look at this @HerHde , I took code to Get room creator session from ```executeAndHandleMissingMember```.
And I send event with him,
I think this is ok for this MR
Tell we what you think of it

**Comment by HerHde** [comment]
I modified it a bit to fix linter errors and removed most of the user detection part to keep it simple, as you pulled the ID from Synapse, anyhow. Further modifications or optimisations can work on this, but for now it is merged. Congratulations and thank you very much!

**Pull request merged**